### PR TITLE
make file descriptors private to the transport

### PR DESCRIPTION
**Pull request by jlevon**
General code has no business knowing about the socket file descriptors.
vfu_attach_ctx() is changed to not return the file descriptor; we'll re-expose a
suitable file descriptor in a follow-up

### CODE REVIEW

File: lib/tran_sock.c [hunk 1]
```diff
@@ -675,30 +695,31 @@ negotiate(vfu_ctx_t *vfu_ctx, int sock)
 static int
 tran_sock_attach(vfu_ctx_t *vfu_ctx)
 {
+    tran_sock_t *ts = vfu_ctx->tran_data;
```
Comment (tmakatos) [inline lib/tran_sock.c:698]: asserting `vfu_ctx` and `ts`

File: lib/tran_sock.c [hunk 2]
```diff
@@ -675,30 +695,31 @@ negotiate(vfu_ctx_t *vfu_ctx, int sock)
...
-    vfu_ctx->conn_fd = conn_fd;
-    return conn_fd;
+    return 0;
 }
 
 static int
 tran_sock_get_request(vfu_ctx_t *vfu_ctx, struct vfio_user_header *hdr,
                       int *fds, size_t *nr_fds)
 {
+    tran_sock_t *ts = vfu_ctx->tran_data;
```
Comment (tmakatos) [inline lib/tran_sock.c:722]: assert `vfu_ctx` and `ts`

File: lib/tran_sock.c [hunk 3]
```diff
@@ -709,14 +730,15 @@ tran_sock_get_request(vfu_ctx_t *vfu_ctx, struct vfio_user_header *hdr,
...
-    return get_msg(hdr, sizeof *hdr, fds, nr_fds, vfu_ctx->conn_fd, sock_flags);
+    return get_msg(hdr, sizeof *hdr, fds, nr_fds, ts->conn_fd, sock_flags);
 }
 
 static int
 tran_sock_recv_body(vfu_ctx_t *vfu_ctx, const struct vfio_user_header *hdr,
                     void **datap)
 {
     size_t body_size = hdr->msg_size - sizeof (*hdr);
+    tran_sock_t *ts = vfu_ctx->tran_data;
```
Comment (tmakatos) [inline lib/tran_sock.c:741]: assert `vfu_ctx`, `hdr`, and `ts`

File: lib/tran_sock.c [hunk 4]
```diff
@@ -750,8 +773,10 @@ tran_sock_reply(vfu_ctx_t *vfu_ctx, uint16_t msg_id,
                 struct iovec *iovecs, size_t nr_iovecs,
                 int *fds, int count, int err)
 {
+    tran_sock_t *ts = vfu_ctx->tran_data;
```
Comment (tmakatos) [inline lib/tran_sock.c:776]: assert `vfu_ctx` and `ts`

File: lib/tran_sock.c [hunk 5]
```diff
@@ -762,26 +787,35 @@ tran_sock_send_msg(vfu_ctx_t *vfu_ctx, uint16_t msg_id,
               struct vfio_user_header *hdr,
               void *recv_data, size_t recv_len)
 {
-    return tran_sock_msg(vfu_ctx->conn_fd, msg_id, cmd, send_data, send_len,
+    tran_sock_t *ts = vfu_ctx->tran_data;
```
Comment (tmakatos) [inline lib/tran_sock.c:790]: assert `vfu_ctx` and `ts`

File: lib/tran_sock.c [hunk 6]
```diff
@@ -762,26 +787,35 @@ tran_sock_send_msg(vfu_ctx_t *vfu_ctx, uint16_t msg_id,
...
                          hdr, recv_data, recv_len);
 }
 
 static void
 tran_sock_detach(vfu_ctx_t *vfu_ctx)
 {
-    if (vfu_ctx->conn_fd != -1) {
-        (void) close(vfu_ctx->conn_fd);
-        vfu_ctx->conn_fd = -1;
+    tran_sock_t *ts = vfu_ctx->tran_data;
```
Comment (tmakatos) [inline lib/tran_sock.c:799]: ditto

File: lib/tran_sock.c [hunk 7]
```diff
@@ -762,26 +787,35 @@ tran_sock_send_msg(vfu_ctx_t *vfu_ctx, uint16_t msg_id,
...
+
+    if (ts->conn_fd != -1) {
+        (void) close(ts->conn_fd);
```
Comment (tmakatos) [inline lib/tran_sock.c:802]: AFAIK we should be handling `EINTR` here, for now let's at least log a warning?

File: lib/tran_sock.c [hunk 8]
```diff
@@ -762,26 +787,35 @@ tran_sock_send_msg(vfu_ctx_t *vfu_ctx, uint16_t msg_id,
...
 static void
 tran_sock_fini(vfu_ctx_t *vfu_ctx)
 {
-    if (vfu_ctx->fd != -1) {
-        (void) close(vfu_ctx->fd);
-        vfu_ctx->fd = -1;
+    tran_sock_t *ts = vfu_ctx->tran_data;
+
+    if (ts->listen_fd != -1) {
+        (void) close(ts->listen_fd);
```
Comment (tmakatos) [inline lib/tran_sock.c:813]: log warning?

File: lib/tran_sock.c [hunk 9]
```diff
@@ -762,26 +787,35 @@ tran_sock_send_msg(vfu_ctx_t *vfu_ctx, uint16_t msg_id,
...
     }
 }
 
 static void
 tran_sock_fini(vfu_ctx_t *vfu_ctx)
 {
-    if (vfu_ctx->fd != -1) {
-        (void) close(vfu_ctx->fd);
-        vfu_ctx->fd = -1;
+    tran_sock_t *ts = vfu_ctx->tran_data;
```
Comment (tmakatos) [inline lib/tran_sock.c:810]: ditto re: assert

**Review by tmakatos — APPROVED** [review]
looks good, minor comments about missing assertion and logging in case `close(2)` fails

File: lib/tran_sock.c (hunk 1 above)
Comment (jlevon) [reply lib/tran_sock.c:698]: done

File: lib/tran_sock.c (hunk 2 above)
Comment (jlevon) [reply lib/tran_sock.c:722]: done

File: lib/tran_sock.c (hunk 3 above)
Comment (jlevon) [reply lib/tran_sock.c:741]: done

File: lib/tran_sock.c (hunk 4 above)
Comment (jlevon) [reply lib/tran_sock.c:776]: done

File: lib/tran_sock.c (hunk 5 above)
Comment (jlevon) [reply lib/tran_sock.c:790]: done

File: lib/tran_sock.c (hunk 6 above)
Comment (jlevon) [reply lib/tran_sock.c:799]: done

File: lib/tran_sock.c (hunk 7 above)
Comment (jlevon) [reply lib/tran_sock.c:802]: added FIXMEs for now

File: lib/tran_sock.c (hunk 8 above)
Comment (jlevon) [reply lib/tran_sock.c:813]: ditto

File: lib/tran_sock.c (hunk 9 above)
Comment (jlevon) [reply lib/tran_sock.c:810]: done

**Review by tmakatos — APPROVED** [review]

**Pull request merged**